### PR TITLE
WEAVIATE-369-implement-bm-25-f-searching

### PR DESCRIPTION
**Pull request by donomii**
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/inverted/prop_length_tracker.go [hunk 1]
```diff
@@ -247,6 +247,33 @@ func (t *PropertyLengthTracker) PropertyMean(propName string) (float32, error) {
 	return sum / totalCount, nil
 }
 
+// BM25F
+func (t *PropertyLengthTracker) PropertyCount(propName string) (float32, float32, error) {
```
Comment (aliszka) [inline adapters/repos/db/inverted/prop_length_tracker.go:251]: this is almost the same as PropertyMean. would be nice to extract and reuse common parts

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 2]
```diff
@@ -162,6 +165,232 @@ func (b *BM25Searcher) retrieveScoreAndSortForSingleTerm(ctx context.Context,
...
+
+	for _, propertyWithBoost := range properties {
+		boost := 1
+		var property = propertyWithBoost
+		if strings.Contains(propertyWithBoost, "^") {
+			property = strings.Split(propertyWithBoost, "^")[0]
+			boostStr := strings.Split(propertyWithBoost, "^")[1]
+			boost, _ = strconv.Atoi(boostStr)
+			fmt.Printf("Boosting %s by %d\n", property, boost)
+		}
```
Comment (aliszka) [inline adapters/repos/db/inverted/bm25_searcher.go:340]: is there any validation for boost expression? like multiple `^`? float/negative/non-number values?
would be nice to have some tests for that

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 3]
```diff
@@ -162,6 +165,232 @@ func (b *BM25Searcher) retrieveScoreAndSortForSingleTerm(ctx context.Context,
...
+
+	//Merge the results from different properties
+	ids := b.mergeIdss(idss, objectByIndexID, propNames)
+
+	b.scoreBM25F(ids, properties)
+	return ids, nil
+}
+
+// BM25F combine and score the results from multiple properties
+func (bm *BM25Searcher) scoreBM25F(ids docPointersWithScore, propName []string) error {
```
Comment (aliszka) [inline adapters/repos/db/inverted/bm25_searcher.go:363]: would be nice to have some tests for that logic

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 4]
```diff
@@ -162,6 +165,232 @@ func (b *BM25Searcher) retrieveScoreAndSortForSingleTerm(ctx context.Context,
...
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "resolve doc ids to objects")
+	}
+
+	return objs, scores, nil
+}
+
+
+// BM25F merge the results from multiple properties
+func (b *BM25Searcher) mergeIdss(idLists []docPointersWithScore, objectByIndexID func(index uint64)*storobj.Object ,  propNames []string) docPointersWithScore {
```
Comment (aliszka) [inline adapters/repos/db/inverted/bm25_searcher.go:237]: would be nice to have some tests for that logic

File: usecases/traverser/explorer.go [hunk 5]
```diff
@@ -131,11 +132,11 @@ func (e *Explorer) getClassKeywordBased(ctx context.Context,
...
 
 	if len(params.KeywordRanking.Properties) == 0 {
-		return nil, errors.Errorf("keyword search (bm25) requires exactly one property")
+		log.Print("No properties specified for keyword search (bm25), selecting all properties")
 	}
 
 	if len(params.KeywordRanking.Properties) > 1 {
-		return nil, errors.Errorf("multi-property keyword search (BM25F) not supported yet")
+		log.Printf("experimental multi-property keyword search (BM25F) active")
 	}
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply usecases/traverser/explorer.go:140]: Debugging leftovers?

**Pull request merged**